dart/collision/CollisionGroup.h

```diff
@@ -0,0 +1,248 @@
+  /// Return index-th ShapeFrames registered to this CollisionGroup. Return
+  /// nullptr if the index is out of the range.
+  const dynamics::ShapeFrame* getShapeFrame(size_t index) const;
+
+  /// Return all the ShapeFrames registered to this CollisionGroup
+  const std::vector<const dynamics::ShapeFrame*>& getShapeFrames() const;
+
+  /// Perform collision detection within this CollisionGroup.
+  bool detect(const Option& option, Result& result);
+
+  /// Perform collision detection with other CollisionGroup.
+  ///
+  /// Return false if the engine type of the other CollisionGroup is different
+  /// from this CollisionObject engine.
+  bool detect(CollisionGroup* group, const Option& option, Result& result);
+
+  /// Return all the CollisionObjects in this CollisionGroup
+  const std::vector<std::shared_ptr<CollisionObject> >& getCollisionObjects();
+
+protected:
+
+  /// Update engine data. This function should be called before the collision
+  /// detection is performed by the engine in most cases.
+  void updateEngineData();
+
+  /// Initialize the collision detection engine data such as broadphase
+  /// algorithm. This function will be called whenever ShapeFrame is either
+  /// added to or removed from this CollisionGroup.
+  virtual void initializeEngineData() = 0;
+
+  /// Notify that a CollisionObject is added so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectAdded(CollisionObject* object) = 0;
+
+  /// Notify that CollisionObjects are added so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectsAdded(
+      const std::vector<CollisionObject*>& collObjects) = 0;
+
+  /// Notify that a CollisionObject is removed so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectRemoved(CollisionObject* object) = 0;
+
+  /// Notify that all the CollisionObjects are remove so that the collision
+  /// detection engine do some relevant work
+  virtual void notifyAllCollisionObjectsRemoved() = 0;
+
+  /// Update the collision detection engine data such as broadphase algorithm.
+  /// This function will be called ahead of every collision checking.
+  virtual void updateCollisionGroupEngineData() = 0;
+
+private:
+
+  /// Collision detector
+  CollisionDetectorPtr mCollisionDetector;
+  // CollisionGroup shares the ownership of CollisionDetector with other
+  // CollisionGroups created from the same CollisionDetector so that the
+  // CollisionDetector doesn't get destroyed as long as at least one
+  // CollisionGroup is alive.
+
+  /// ShapeFrames added to this CollisionGroup
+  std::vector<const dynamics::ShapeFrame*> mShapeFrames;
+
+  /// CollisionObjects associated with the added ShapeFrames
+  std::vector<std::shared_ptr<CollisionObject>> mCollisionObjects;
+  // CollisionGroup also shares the ownership of CollisionObjects across other
+  // CollisionGroups for the same reason with above.
```

Comment:
We still need to get `vector<CollisionObject>` from `map` by the way. 

I think we should have a consistent way of getting an iterable object from stl containers as I suggested #613. This is similar problem to getting `vector<const Object*>` from `vector<Object*>`. We create a temporary `vector<const Object*>` and return it without performance consideration in this case.